src/index/ivf/ivf.cc

```diff
@@ -270,73 +272,153 @@ IvfIndexNode<T>::Train(const DataSet& dataset, const Config& cfg) {
+
+            // create quantizer for the training
+            auto qzr = use_elkan ? 
+                std::make_unique<faiss::IndexFlatElkan>(dim, metric.value()) :
+                std::make_unique<faiss::IndexFlat>(dim, metric.value());
+            // create index. Index does not own qzr
+            index = std::make_unique<faiss::IndexIVFFlat>(qzr.get(), dim, nlist, metric.value(), is_cosine);
+            // train
             index->train(rows, (const float*)data);
+            // replace quantizer if needed
+            if (use_elkan) {
+                qzr = to_index_flat(std::move(qzr));
+                index->quantizer = qzr.get();
+            }
+            // transfer ownership of qzr to index
+            qzr.release();
+            index->own_fields = true;
         }
         if constexpr (std::is_same<faiss::IndexIVFFlatCC, T>::value) {
             const IvfFlatCcConfig& ivf_flat_cc_cfg = static_cast<const IvfFlatCcConfig&>(cfg);
             auto nlist = MatchNlist(rows, ivf_flat_cc_cfg.nlist.value());
-            qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = std::make_unique<faiss::IndexIVFFlatCC>(qzr, dim, nlist, ivf_flat_cc_cfg.ssize.value(),
+
+            const bool use_elkan = ivf_flat_cc_cfg.use_elkan.value_or(true);
+
+            // create quantizer for the training
+            auto qzr = use_elkan ? 
+                std::make_unique<faiss::IndexFlatElkan>(dim, metric.value()) :
+                std::make_unique<faiss::IndexFlat>(dim, metric.value());
+            // create index. Index does not own qzr
+            index = std::make_unique<faiss::IndexIVFFlatCC>(qzr.get(), dim, nlist, ivf_flat_cc_cfg.ssize.value(),
                                                             metric.value(), is_cosine);
+            // train
             index->train(rows, (const float*)data);
+            // replace quantizer if needed
+            if (use_elkan) {
+                qzr = to_index_flat(std::move(qzr));
+                index->quantizer = qzr.get();
+            }
+            // transfer ownership of qzr to index
+            qzr.release();
+            index->own_fields = true;
             // ivfflat_cc has no serialize stage, make map at build stage
             index->make_direct_map(true, faiss::DirectMap::ConcurrentArray);
         }
         if constexpr (std::is_same<faiss::IndexIVFPQ, T>::value) {
             const IvfPqConfig& ivf_pq_cfg = static_cast<const IvfPqConfig&>(cfg);
             auto nlist = MatchNlist(rows, ivf_pq_cfg.nlist.value());
             auto nbits = MatchNbits(rows, ivf_pq_cfg.nbits.value());
-            qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = std::make_unique<faiss::IndexIVFPQ>(qzr, dim, nlist, ivf_pq_cfg.m.value(), nbits, metric.value());
+
+            const bool use_elkan = ivf_pq_cfg.use_elkan.value_or(true);
```

Comment:
Also, so much dup codes here (I know it is like this before this PR). Any possibility to concentrate them?

Reply:
Refactored `IndexFlatElkan` a bit.